### PR TITLE
Expanded support for CXSMILES features

### PR DESCRIPTION
**Pull request by greglandrum**
Adds support for parsing:
- `rb`  : ring-bond count  - converts to query
- `u` : unsaturation - converts to query
- `s` : non-hydrogen substitution count - converts to query
- `m`: variable attachment points  - adds bond property corresponding to what you get from a Mol file
- `LN`: linknodes - adds mol property corresponding to what you get from a Mol file.

It also adds a new SMARTS primitive, `d`, which queries for the non-hydrogen degree of an atom, because that seems useful. There's an argument that this should go in its own PR, and I'm happy to do that if others prefer.

Other things in this PR:
- extended the documentation of which pieces of CXSMILES are supported
- moves a couple useful functions from `FileParserUtils.h` to `QueryOps.h` and deprecates the old versions
- remove an unused variable from `AdjustQuery.cpp`
- correct a bug in `queryAtomNonHydrogenDegree()`
- additional cleanup in `QueryOps.h`

### CODE REVIEW

**Comment by bp-kelley** [comment]
Most of the way through.  I can't say I understand the Link nodes though, I'll have to look up the CXSmiles docs.

**Comment by greglandrum** [comment]
> Most of the way through. I can't say I understand the Link nodes though, I'll have to look up the CXSmiles docs.

Yeah, those took a while to figure out. I also knew that I'd have to do it again every time I came back to the code, which is why I added the long comment/explanation here:
https://github.com/greglandrum/rdkit/blob/fix/cxsmiles_mdl_queries/Code/GraphMol/SmilesParse/CXSmilesOps.cpp#L343

File: Code/GraphMol/QueryOps.h [hunk 1]
```diff
@@ -80,19 +80,27 @@ static inline int queryAtomExplicitDegree(Atom const *at) {
 static inline int queryAtomTotalDegree(Atom const *at) {
   return at->getTotalDegree();
 };
+//! includes D and T
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/QueryOps.h:83]: exactly. The query here returns the number of bonds to things that are not H, and D and T are not considered to be H for those purposes. I will clarify that

File: Code/GraphMol/QueryOps.h [hunk 2]
```diff
@@ -80,19 +80,27 @@ static inline int queryAtomExplicitDegree(Atom const *at) {
...
+       boost::make_iterator_range(at->getOwningMol().getAtomNeighbors(at))) {
+    const auto nbr = at->getOwningMol()[nbri];
+    if (nbr->getAtomicNum() != 1 || nbr->getIsotope() > 1) {
+      res++;
+    }
+  }
+
+  return res;
+};
 static inline int queryAtomHeavyAtomDegree(Atom const *at) {
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/QueryOps.h:96]: not as defined here.

File: Code/GraphMol/SmilesParse/CXSmilesOps.cpp [hunk 3]
```diff
@@ -250,6 +250,249 @@ bool parse_coordinate_bonds(Iterator &first, Iterator last, RDKit::RWMol &mol) {
...
+  return true;
+}
+
+template <typename Iterator>
+bool parse_linknodes(Iterator &first, Iterator last, RDKit::RWMol &mol) {
+  // these look like: |LN:1:1.3.2.6,4:1.4.3.6|
+  // that's two records:
+  //   1:1.3.2.6: 1-3 repeats, atom 1-2, 1-6
+  //   4:1.4.3.6: 1-4 repeats, atom 4-4, 4-6
+  // which maps to the property value "1 3 2 2 3 2 7|1 4 2 5 4 5 7"
```
Comment (ricrogz) [inline Code/GraphMol/SmilesParse/CXSmilesOps.cpp:347]: I checked the [CXSMILES docs for link nodes](https://docs.chemaxon.com/display/docs/ChemAxon_Extended_SMILES_and_SMARTS_-_CXSMILES_and_CXSMARTS.html), and it says this:

> If the link node has only two connections, then the first and second outer atom indexes are obvious, so they are omitted.

So, apparently, in some cases we can have link nodes with just two "parameters", like `|LN:1:1.3|`. Should we support these too?

Comment (greglandrum) [reply Code/GraphMol/SmilesParse/CXSmilesOps.cpp:347]: In principle yes, but when I was writing the code I couldn't figure out what that would actually mean and couldn't make marvin generate an example. Now I think I get it and I will update the code and add some tests.

Comment (ricrogz) [reply Code/GraphMol/SmilesParse/CXSmilesOps.cpp:347]: Makes sense.

**Pull request merged**